contracts/EpochTokenLocker.sol

```diff
@@ -0,0 +1,146 @@
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount = balanceStates[msg.sender][token].pendingDeposits.amount.add(amount);
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function requestWithdraw(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit withdrawRequest(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdraw(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token); // withdrawn amount might just be deposited before
+        
+        require(
+            balanceStates[msg.sender][token].pendingWithdraws.stateIndex < currentStateIndex,
+            "withdraw was not registered previously"
+        );
+
+        require(
+            balanceStates[msg.sender][token].pendingWithdraws.amount >= amount,
+            "registered withdraw-amount was not sufficient"
+        );
+
+        require(
+            balanceStates[msg.sender][token].balance >= amount,
+            "balances not sufficient"
+        );
+
+        balanceStates[msg.sender][token].balance = balanceStates[msg.sender][token].balance.sub(amount);
+        delete balanceStates[msg.sender][token].pendingWithdraws;
+
+        ERC20(token).transfer(msg.sender, amount);
+        emit Withdraw(msg.sender, token, amount);
+    }
+
+    function updateDepositsBalance(address user, address token) public {
+        if (balanceStates[user][token].pendingDeposits.stateIndex < currentStateIndex) {
+            balanceStates[user][token].balance = balanceStates[user][token].balance.add(
+                balanceStates[user][token].pendingDeposits.amount
+            );
+
+            delete balanceStates[user][token].pendingDeposits;
+        }
+    }
+
+    /**
+     * view functions
+     */
+    function getPendingDepositAmount(address user, address token) public view returns(uint) {
+        return balanceStates[user][token].pendingDeposits.amount;
+    }
+
+    function getPendingDepositBatchNumber(address user, address token) public view returns(uint) {
+        return balanceStates[user][token].pendingDeposits.stateIndex;
+    }
+
+    function getPendingWithdrawAmount(address user, address token) public view returns(uint) {
+        return balanceStates[user][token].pendingWithdraws.amount;
+    }
+
+    function getPendingWithdrawBatchNumber(address user, address token) public view returns(uint) {
+        return balanceStates[user][token].pendingWithdraws.stateIndex;
+    }
+
+    function getBalance(address user, address token) public view returns(uint) {
+        return balanceStates[user][token].balance;
+    }
+
+    /**
+     * internal functions
+     */
+    function addBalance(address user, address token, uint amount) internal {
```

Comment:
I don't see this function used in the smart contract. I see that is used only for testing purposes.

If you need it only for testing, is not better to do one of the following?:
- define it just in the test contracts
- Use the mocking library
- Do a setup where u do the actual deposit of tokens

Reply:
The function defines an interface, which will be used later. At least to my current thinking

Reply:
Should it then really be defined as an interface? At the moment it's just a one-off internal functions with no call-sites (except for some contract used for testing). Do we need it for this PR?